service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,454 @@
+            certFromDb.setName(name);
+        }
+        String description = giftCertificatePatchDto.getDescription();
+        if (description != null) {
+            certFromDb.setDescription(description);
+        }
+        BigDecimal price = giftCertificatePatchDto.getPrice();
+        if (price != null) {
+            certFromDb.setPrice(price);
+        }
+        Integer duration = giftCertificatePatchDto.getDuration();
+        if (duration != null) {
+            certFromDb.setDuration(duration);
+        }
+
+        String currentTime = DateProvider.getNowAsString();
+        certFromDb.setLastUpdateDate(currentTime);                  //init dto
+
+        giftCertificateDao.update(certFromDb, certificateId);
+
+        List<TagDto> passedTags = giftCertificatePatchDto.getTags();  //check passed tags
+        List<Tag> certTags = new ArrayList<>(tagDao.getAllByCertificateId(certificateId));
+
+        if (passedTags != null) {
+            passedTags.forEach(tagDto -> {
+                String tagDtoName = tagDto.getName();
+                Optional<Tag> tagFromDbOpt = tagDao.findByName(tagDtoName); //check if the tag with such name in db
+
+                if (tagFromDbOpt.isPresent()) {
+                    Tag tagFromDb = tagFromDbOpt.get();
+                    Long tagId = tagFromDb.getId();
+                    boolean isTagAlreadyAttachedToCertificate = certTags.stream()
+                            .map(Tag::getId)
+                            .anyMatch(tagId::equals);
+                    if (!isTagAlreadyAttachedToCertificate) {
+                        certificateTagsDao.save(certificateId, tagFromDb.getId());
+                        certTags.add(tagFromDb);
+                    }
+                } else {
+                    Tag tagEntity = modelMapper.map(tagDto, Tag.class);
+                    Tag savedTag = tagDao.save(tagEntity);
+                    certificateTagsDao.save(certificateId, savedTag.getId());
+                    certTags.add(savedTag);
+                }
+            });
+        }
+
+        List<TagDto> tagsDto = certTags.stream()
+                .map(tag -> modelMapper.map(tag, TagDto.class))
+                .collect(Collectors.toList());
+        GiftCertificateDto dto = modelMapper.map(certFromDb, GiftCertificateDto.class);
+        dto.setTags(tagsDto);
+
+        return dto;
+    }
+
+    /**
+     * This method get all GiftCertificate entities from database and converts them to GiftCertificateDto.
+     *
+     * @return list of all GiftCertificateDto.
+     * @throws GiftCertificateNotFoundException if there are no GiftCertificate entities in db.
+     * @since 1.0
+     */
+    private List<GiftCertificateDto> getAll() {
+        List<GiftCertificate> entities = giftCertificateDao.findAll();
+        if (entities.isEmpty()) {
+            throw new GiftCertificateNotFoundException("There are no gift certificates in DB");
+        }
+
+        return entities.stream()
```

Comment:
I see this part everywhere in get... methods, please extract it if possible

Reply:
Remove all getAll*() methods. Replaced it by Criterias. Servica layer requests concrete Criteria with concrete Parameters(fields for searching) from CriteriaFactory and passes it to Dao layer. Sorting now is a part of Service layer and uses Comparators instead of ORDER BY sql queries.

commit cb8a000